File: engines/unity/flutter_plugin/ios/cytoid_game_core/Sources/cytoid_game_core/RuntimeStateMachine.swift
import Foundation

/// Native runtime lifecycle state machine shared by `CytoidGameCoreBridge` and
/// `MockGameCoreBridge`. Encodes the v2 host protocol transitions:
///
///   unavailable → starting          (host requested startup)
///   starting    → ready             (engine.ready)
///   ready       → busy              (session.started; sets activeSessionId)
///   busy        → ready             (session.result; clears activeSessionId)
///   ready|busy  → suspended         (app backgrounded; saves prior state)
///   suspended   → prior state       (resume; single-slot memory, no stack)
///   any         → failed            (unrecoverable; sets lastError)
///
/// Pause during `starting` stays `starting` — no transition.
///
/// The machine owns only the lifecycle bookkeeping. Downstream tasks (T4/T5/T6)
/// are responsible for synthesising `session.failed` when `onFailure` fires
/// with a non-null `activeSessionId`; see `.omo/notepads/v2-host-impl/decisions.md`.
public final class RuntimeStateMachine {
    public private(set) var state: RuntimeState = .unavailable
    public private(set) var generation: Int = 0
    public private(set) var activeSessionId: String?
    public private(set) var lastError: GameCoreError?

    // Single-slot resume memory. NOT a stack. Cleared on resume or failure.
    private var priorStateForResume: RuntimeState?

    public init() {}

    /// v2 runtime snapshot, with conditional optionality matching the spec:
    /// required keys `engine`, `mode`, `state`, `generation` always present;
    /// `activeSessionId` only when `state = busy`; `error` only when
    /// `state = failed`.
    public func snapshot(engine: String, mode: String) -> [String: Any] {
        var map: [String: Any] = [
            "engine": engine,
            "mode": mode,
            "state": state.wireName,
            "generation": generation,
        ]
        if state == .busy {
            // busy implies activeSessionId != nil by construction — the only
            // path into busy is onSessionStarted(id), which sets both.
            guard let activeSessionId else {
                preconditionFailure("RuntimeStateMachine: busy state requires activeSessionId")
            }
            map["activeSessionId"] = activeSessionId
        }
        if state == .failed, let lastError {
            map["error"] = lastError.toMap()
        }
        return map
    }

    /// unavailable → starting. Idempotent if already past unavailable.
    public func onRequestStart() {
        if state == .unavailable {
            state = .starting
        }
    }

    /// starting → ready (or failed → ready on recovery). Bumps `generation`
    /// when entering ready, per v2 § Active-Session Runtime Failure ("next
    /// engine.ready MUST carry an incremented generation"). No-op from any
    /// other state: a duplicate or late ready arriving while busy or
    /// suspended must not reset state and corrupt the active session or the
    /// suspend invariant.
    public func onEngineReady() {
        if state != .starting && state != .failed {
            return
        }
        generation += 1
        state = .ready
        lastError = nil
    }

    /// ready → busy. Sets `activeSessionId`. No-op outside ready.
    public func onSessionStarted(sessionId: String) {
        if state == .ready {
            activeSessionId = sessionId
            state = .busy
        }
    }

    /// busy → ready. Clears `activeSessionId`. No-op outside busy/suspended.
    public func onSessionEnded() {
        switch state {
        case .busy:
            activeSessionId = nil
            state = .ready
        case .suspended:
            // Session ended while suspended (cancel/result during
            // backgrounding): clear the active session and neutralize the
            // resume target so onResume lands in ready, not stale busy.
            activeSessionId = nil
            priorStateForResume = .ready
        default:
            break
        }
    }

    /// ready|busy → suspended. Saves the prior state in `priorStateForResume`
    /// so `onResume` can restore it. Pause during starting stays starting —
    /// no transition, no resume slot.
    public func onSuspend() {
        switch state {
        case .ready, .busy:
            priorStateForResume = state
            state = .suspended
        case .unavailable, .starting, .suspended, .failed:
            break
        }
    }

    /// suspended → prior state (single-slot memory). No-op outside suspended.
    public func onResume() {
        guard state == .suspended, let prior = priorStateForResume else { return }
        state = prior
        priorStateForResume = nil
    }

    /// any → failed. Sets `lastError`, clears `activeSessionId` and the resume
    /// slot. Active-session failure routing (whether to emit `session.result`
    /// vs `engine.error`) is decided by the bridge based on the pre-failure
    /// value of `activeSessionId`; see `decisions.md` T3 entry.
    public func onFailure(error: GameCoreError) {
        lastError = error
        activeSessionId = nil
        priorStateForResume = nil
        state = .failed
    }

    /// Test-only: directly clear `activeSessionId` without changing state.
    /// Used by the bridge when a suspended session is cancelled by backgrounding
    /// (per T3 lifecycle rule: clear on suspended only if cancelled). Does NOT
    /// affect state — the runtime stays suspended until resumed.
    public func clearActiveSessionForCancellation() {
        activeSessionId = nil
    }
}

File: engines/unity/flutter_plugin/android/src/main/kotlin/org/cytoid/gamecore/RuntimeStateMachine.kt
package org.cytoid.gamecore

import androidx.annotation.VisibleForTesting

/**
 * Native runtime lifecycle state machine shared by [CytoidGameCoreBridge] and
 * [MockGameCoreBridge]. Encodes the v2 host protocol transitions:
 *
 *   UNAVAILABLE → STARTING            (host requested startup)
 *   STARTING    → READY               (engine.ready)
 *   READY       → BUSY                (session.started; sets activeSessionId)
 *   BUSY        → READY               (session.result; clears activeSessionId)
 *   READY|BUSY  → SUSPENDED           (app backgrounded; saves prior state)
 *   SUSPENDED   → prior state         (resume; single-slot memory, no stack)
 *   any         → FAILED              (unrecoverable; sets lastError)
 *
 * Pause during STARTING stays STARTING — no transition.
 *
 * The machine owns only the lifecycle bookkeeping. Downstream tasks (T4/T5/T6)
 * are responsible for synthesising `session.failed` when [onFailure] fires
 * with a non-null [activeSessionId]; see `.omo/notepads/v2-host-impl/decisions.md`.
 */
class RuntimeStateMachine {
    @Volatile
    var state: RuntimeState = RuntimeState.UNAVAILABLE
        private set

    @Volatile
    var generation: Int = 0
        private set

    @Volatile
    var activeSessionId: String? = null
        private set

    @Volatile
    var lastError: GameCoreError? = null
        private set

    // Single-slot resume memory. NOT a stack. Cleared on resume or failure.
    @Volatile
    private var priorStateForResume: RuntimeState? = null

    /**
     * v2 runtime snapshot, with conditional optionality matching the spec:
     * required keys `engine`, `mode`, `state`, `generation` always present;
     * `activeSessionId` only when `state = busy`; `error` only when
     * `state = failed`.
     */
    fun snapshot(engine: String, mode: String): Map<String, Any?> = buildMap {
        put("engine", engine)
        put("mode", mode)
        put("state", state.wireName)
        put("generation", generation)
        if (state == RuntimeState.BUSY) {
            // BUSY implies activeSessionId != null by construction — the only
            // path into BUSY is onSessionStarted(id), which sets both.
            put("activeSessionId", checkNotNull(activeSessionId) {
                "RuntimeStateMachine: BUSY state requires activeSessionId"
            })
        }
        if (state == RuntimeState.FAILED) {
            lastError?.let { put("error", it.toMap()) }
        }
    }

    /** UNAVAILABLE → STARTING. Idempotent if already past UNAVAILABLE. */
    @VisibleForTesting
    fun onRequestStart() {
        if (state == RuntimeState.UNAVAILABLE) {
            state = RuntimeState.STARTING
        }
    }

    /**
     * STARTING → READY (or FAILED → READY on recovery). Bumps [generation]
     * when entering READY, per v2 § Active-Session Runtime Failure ("next
     * engine.ready MUST carry an incremented generation"). No-op from any
     * other state: a duplicate or late ready arriving while BUSY or
     * SUSPENDED must not reset state and corrupt the active session or the
     * suspend invariant.
     */
    @VisibleForTesting
    fun onEngineReady() {
        if (state != RuntimeState.STARTING && state != RuntimeState.FAILED) {
            return
        }
        generation += 1
        state = RuntimeState.READY
        lastError = null
    }

    /** READY → BUSY. Sets [activeSessionId]. No-op outside READY. */
    @VisibleForTesting
    fun onSessionStarted(sessionId: String) {
        if (state == RuntimeState.READY) {
            activeSessionId = sessionId
            state = RuntimeState.BUSY
        }
    }

    /** BUSY → READY. Clears [activeSessionId]. No-op outside BUSY/SUSPENDED. */
    @VisibleForTesting
    fun onSessionEnded() {
        when (state) {
            RuntimeState.BUSY -> {
                activeSessionId = null
                state = RuntimeState.READY
            }
            RuntimeState.SUSPENDED -> {
                // Session ended while suspended (e.g. session.cancel / session.result
                // arrived during backgrounding): clear the active session and
                // neutralize the resume target so onResume lands in READY, not
                // stale BUSY with a dead sessionId.
                activeSessionId = null
                priorStateForResume = RuntimeState.READY
            }
            else -> Unit
        }
    }

    /**
     * READY|BUSY → SUSPENDED. Saves the prior state in [priorStateForResume]
     * so [onResume] can restore it. Pause during STARTING stays STARTING —
     * no transition, no resume slot.
     */
    @VisibleForTesting
    fun onSuspend() {
        when (state) {
            RuntimeState.READY, RuntimeState.BUSY -> {
                priorStateForResume = state
                state = RuntimeState.SUSPENDED
            }
            else -> Unit
        }
    }

    /** SUSPENDED → prior state (single-slot memory). No-op outside SUSPENDED. */
    @VisibleForTesting
    fun onResume() {
        val prior = priorStateForResume
        if (state == RuntimeState.SUSPENDED && prior != null) {
            state = prior
            priorStateForResume = null
        }
    }

    /**
     * any → FAILED. Sets [lastError], clears [activeSessionId] and the resume
     * slot. Active-session failure routing (whether to emit `session.result`
     * vs `engine.error`) is decided by the bridge based on the pre-failure
     * value of [activeSessionId]; see `decisions.md` T3 entry.
     */
    @VisibleForTesting
    fun onFailure(error: GameCoreError) {
        lastError = error
        activeSessionId = null
        priorStateForResume = null
        state = RuntimeState.FAILED
    }

    /**
     * Test-only: directly clear [activeSessionId] without changing state.
     * Used by the bridge when a SUSPENDED session is cancelled by backgrounding
     * (per T3 lifecycle rule: clear on SUSPENDED only if cancelled). Does NOT
     * affect state — the runtime stays SUSPENDED until resumed.
     */
    @VisibleForTesting
    fun clearActiveSessionForCancellation() {
        activeSessionId = null
    }

    /**
     * Reset to [RuntimeState.UNAVAILABLE]. Called when the runtime surface is
     * destroyed without an active session (e.g. the OS reclaimed the Unity
     * Activity). The host MUST call `startRuntime()` again before launching a
     * new session.
     *
     * Does NOT reset [generation]: the next `onEngineReady` bumps it naturally
     * from STARTING. [activeSessionId] is already null by contract (the caller
     * only invokes this when no session is active). [lastError] and the resume
     * slot are cleared to give the host a clean slate on restart.
     */
    @VisibleForTesting
    fun reset() {
        state = RuntimeState.UNAVAILABLE
        lastError = null
        priorStateForResume = null
    }
}
